x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,177 @@
+	poolIdLeaving = gammtypes.MustGetPoolIdFromShareDenom(sharesToMigrate.Denom)
+
+	// Ensure a governance sanctioned link exists between the balancer pool and the concentrated pool.
+	poolIdEntering, err = k.gk.GetLinkedConcentratedPoolID(ctx, poolIdLeaving)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Get the concentrated pool from the provided ID and type cast it to ConcentratedPoolExtension.
+	concentratedPool, err := k.clk.GetPoolFromPoolIdAndConvertToConcentrated(ctx, poolIdEntering)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// Check that lockID corresponds to sender, and contains correct denomination of LP shares.
+	lock, err := k.validateLockForUnpool(ctx, sender, poolIdLeaving, lockId)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+	gammSharesInLock := lock.Coins[0]
+	preUnlockLock := *lock
+
+	// Before we break the lock, we must note the time remaining on the lock.
+	remainingLockTime := k.getExistingLockRemainingDuration(ctx, lock)
+
+	// Check if the lock has a corresponding synthetic lock.
+	synthLockBeforeMigration := k.lk.GetAllSyntheticLockupsByLockup(ctx, lockId)
+
+	// If it does, check if it is superfluid delegated or undelegating.
+	wasSuperfluidUndelegatingBeforeMigration := len(synthLockBeforeMigration) > 0 && strings.Contains(synthLockBeforeMigration[0].SynthDenom, "superunbonding")
+	wasSuperfluidDelegatedBeforeMigration := len(synthLockBeforeMigration) > 0 && strings.Contains(synthLockBeforeMigration[0].SynthDenom, "superbonding")
+
+	// If it is superfluid delegated or undelegating, get the validator address from the synth denom.
+	valAddr := ""
+	if wasSuperfluidDelegatedBeforeMigration || wasSuperfluidUndelegatingBeforeMigration {
+		valAddr = strings.Split(synthLockBeforeMigration[0].SynthDenom, "/")[4]
+	}
+
+	// If the lock was superfluid delegated, superfluid undelegate it
+	// This deletes the connection between the lock and the intermediate account, deletes the synthetic lock, and burns the synthetic osmo.
+	intermediateAccount := types.SuperfluidIntermediaryAccount{}
+	if wasSuperfluidDelegatedBeforeMigration {
+		// superfluid undelegate and break any underlying synthetic locks
+		// this is the same as SuperfluidUndelegate, but does not create a corresponding superunbonding synthetic lock
+		intermediateAccount, err = k.SuperfluidUndelegateToConcentratedPosition(ctx, sender.String(), lockId)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+		}
+	}
+
+	// Finish unlocking directly for locked or unlocking locks
+	err = k.lk.ForceUnlock(ctx, *lock)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
+	}
+
+	// If shares to migrate is not specified, we migrate all shares.
+	if sharesToMigrate.IsZero() {
+		sharesToMigrate = gammSharesInLock
+	}
```

Comment:
Do you think it might make sense to force users to provide this value and error out if not given?

Reply:
The reason I made it this way was simplicity on the frontend, if a user wants to migrate everything then the frontend doesn't have to determine their current total liquidity. But I can change this if you think it should be the way you said in the comment